archive/100-theorems-list/93_birthday_problem.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import data.fintype.card
+import data.nat.factorial
+import tactic
+
+/-!
+# Birthday Problem
+
+This file proves Theorem 93 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/).
+
+As of writing, probability is not implemented in Lean, so we instead state the birthday problem
+in terms of injective functions. We instead prove a general result about ‖α ↪ β‖.
+-/
+
+open_locale classical nat
+
+open finset function nat
+local notation `|` x `|` := finset.card x
+local notation `‖` x `‖` := fintype.card x
+
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation. -/
+def desc_fac (n : ℕ) : ℕ → ℕ
+| 0 := 1
+| (k + 1) := (n + k + 1) * desc_fac k
+
+@[simp] lemma desc_fac_zero (n : ℕ) : desc_fac n 0 = 1 := rfl
+
+@[simp] lemma zero_desc_fac (k : ℕ) : desc_fac 0 k = k! :=
+begin
+  induction k with t ht, refl,
+  unfold desc_fac, rw [ht, zero_add, factorial_succ]
+end
+
+lemma desc_fac_succ {n k : ℕ} : desc_fac n k.succ = (n + k + 1) * desc_fac n k := rfl
+
+lemma succ_desc_fac {n k : ℕ} : (n + 1) * desc_fac n.succ k = (n + k + 1) * desc_fac n k :=
+begin
+  induction k with t ht, simp!, rw desc_fac_succ, rw desc_fac_succ,
+  have : (n + 1) * ((n.succ + t + 1) * desc_fac n.succ t)
+       = (n.succ + t + 1) * ((n + 1) * desc_fac n.succ t), by ac_refl,
+  rw this, rw ht, repeat {rw succ_eq_add_one}, ring
+end
+
+/-- Prove that `desc_fac` is what it is promised to be. Stated divison-less for ease. -/
+theorem eval_desc_fac (n : ℕ) : ∀ k : ℕ, (n + k)! = n! * desc_fac n k
+| 0 := by simp!
+| (k + 1) := by unfold desc_fac; rw [←mul_assoc, mul_comm n!, mul_assoc, ←eval_desc_fac]; simp!
+
+def embedding_of_subtype (α β) [fintype α] [fintype β] : (α ↪ β) ≃ {f : α → β // injective f} :=
+{ to_fun := λ f, ⟨f, f.injective⟩,
+  inv_fun := λ f, ⟨f.val, f.property⟩,
+  left_inv := λ f, by {ext, simp},
+  right_inv := λ f, by simp }
+
+-- `decidable_pred (@injective α β)` and various variations didn't give me an instance 🤷‍♂️
+noncomputable instance fintype.embedding {α β} [fintype α] [fintype β] : fintype (α ↪ β) :=
+fintype.of_equiv {f : α → β // injective f} (embedding_of_subtype α β).symm
+
+/-- Establishes the cardinality of the type of injective functions `fin n ↪ β`. -/
+lemma card_inj' (n : ℕ) (β) [fintype β] (h : n ≤ ‖β‖) : ‖fin n ↪ β‖ = desc_fac (‖β‖ - n) n :=
+begin
+  induction n with n hn,
+    rw [desc_fac_zero], nontriviality (fin 0 ↪ β),
+    obtain ⟨f, g, ne⟩ := exists_pair_ne (fin 0 ↪ β),
+    exfalso, apply ne, ext x, exact x.elim0,
+
+  let extend : (fin n → β) → β → fin n.succ → β :=
+    λ f : fin n → β, λ b : β, fin.cons b f,
+
+  let equiv_classes : (fin n ↪ β) → finset (fin n.succ ↪ β) :=
+    λ f : fin n ↪ β, univ.filter (λ g : fin n.succ ↪ β, ∃ k : β, extend f k = g),
```

Comment:
```suggestion
    λ f : fin n ↪ β, univ.filter (λ g : fin n.succ ↪ β, ∃ k : β, fin.cons k f = g),
```
I think you can omit `extend` entirely

Reply:
I tried doing it before, and ran into type-checker issues; the "let" was basically to force the expected type. Tried it out again in Gitpod and it seemed to be the same